// @ciscospark/plugin-metrics/dist/metrics.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _assign = require('babel-runtime/core-js/object/assign');

var _assign2 = _interopRequireDefault(_assign);

var _defaults = require('lodash/defaults');

var _defaults2 = _interopRequireDefault(_defaults);

var _sparkCore = require('@ciscospark/spark-core');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Metrics = _sparkCore.SparkPlugin.extend({
  namespace: 'Metrics',

  submit: function submit(key, value) {
    value = value || {};
    (0, _defaults2.default)(value, {
      appType: this.config.appType,
      env: process.env.NODE_ENV || 'development',
      version: this.spark.version,
      time: Date.now(),
      postTime: Date.now()
    });

    return this.request({
      method: 'POST',
      service: 'metrics',
      resource: 'metrics',
      body: {
        metrics: [(0, _assign2.default)({ key: key }, value)]
      }
    });
  }
}); /**!
     *
     * Copyright (c) 2015-2016 Cisco Systems, Inc. See LICENSE file.
     * @private
     */

exports.default = Metrics;
