// ciscospark/dist/ciscospark.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _defaults = require('lodash/defaults');

var _defaults2 = _interopRequireDefault(_defaults);

var _deleteProperty = require('babel-runtime/core-js/reflect/delete-property');

var _deleteProperty2 = _interopRequireDefault(_deleteProperty);

var _get = require('lodash/get');

var _get2 = _interopRequireDefault(_get);

var _set = require('lodash/set');

var _set2 = _interopRequireDefault(_set);

var _has = require('lodash/has');

var _has2 = _interopRequireDefault(_has);

require('@ciscospark/plugin-phone');

var _sparkCore = require('@ciscospark/spark-core');

var _sparkCore2 = _interopRequireDefault(_sparkCore);

var _auth = require('./interceptors/auth');

var _auth2 = _interopRequireDefault(_auth);

var _memberships = require('./plugins/memberships');

var _memberships2 = _interopRequireDefault(_memberships);

var _messages = require('./plugins/messages');

var _messages2 = _interopRequireDefault(_messages);

var _people = require('./plugins/people');

var _people2 = _interopRequireDefault(_people);

var _rooms = require('./plugins/rooms');

var _rooms2 = _interopRequireDefault(_rooms);

var _teams = require('./plugins/teams');

var _teams2 = _interopRequireDefault(_teams);

var _teamMemberships = require('./plugins/team-memberships');

var _teamMemberships2 = _interopRequireDefault(_teamMemberships);

var _webhooks = require('./plugins/webhooks');

var _webhooks2 = _interopRequireDefault(_webhooks);

var _config = require('./config');

var _config2 = _interopRequireDefault(_config);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

// This shouldn't be necessary once the plugins are moved to their own packages
var interceptors = {}; /**!
                        *
                        * Copyright (c) 2015-2016 Cisco Systems, Inc. See LICENSE file.
                        * @private
                        */

/* eslint camelcase: [0] */

if (!_sparkCore.children.device) {
  interceptors['AuthInterceptor'] = _auth2.default.create;
}

(0, _sparkCore.registerPlugin)('memberships', _memberships2.default, {
  interceptors: interceptors
});
(0, _sparkCore.registerPlugin)('messages', _messages2.default);
(0, _sparkCore.registerPlugin)('people', _people2.default);
(0, _sparkCore.registerPlugin)('rooms', _rooms2.default);
(0, _sparkCore.registerPlugin)('teams', _teams2.default);
(0, _sparkCore.registerPlugin)('teamMemberships', _teamMemberships2.default);
(0, _sparkCore.registerPlugin)('webhooks', _webhooks2.default);

/**
 * SDK for Cisco Spark
 *
 * ## Installation
 *
 * ```javascript
 * npm install --save ciscospark
 * ```
 *
 * ## A Note on Browsers
 * Ciscospark is fully browser compatible but we don't distribute a browserified
 * bundle at this time; you'll need to build the bundle yourself. We use
 * [browserify](http://browserify.org/) internally and
 * [webpack](https://webpack.github.io/) should work as well.
 *
 * ## Getting Started
 *
 *
 * The quickest way to get started is to set your access token as an environment
 * variable:
 *
 * ```javascript
 * <%= gettingstarted__accesstoken %>
 * ```

 * ### Refresh Tokens
 *
 * For long-running use cases, you'll need to provide a refresh token, client
 * id, and client secret:
 *
 * ```javascript
 * TODO
 * ```
 *
 * ## Runtime Configuration
 *
 * While environment variables are handy for development, they don't really help
 * you write an app for lots of users. You can pass credentials to the spark
 * using init.
 *
 * ```javascript
 * TODO
 * ```
 *
 * ## OAuth
 *
 * OAuth is baked right into spark so you don't need to figure it out.
 *
 * To kick off an OAuth login, simply call `spark.authenticate()` which will
 * direct the current app to our login page.
 *
 * ```javascript
 * TODO
 * ```
 *
 * To refresh an access token, call `spark.authorize()`. (Note: this should
 * generally happen for you automatically).
 *
 * ```javascript
 * TODO
 * ```
 *
 * @class CiscoSpark
 * @extends SparkCore
 */
var ciscospark = new _sparkCore2.default({
  credentials: {
    authorization: {
      access_token: process.env.CISCOSPARK_ACCESS_TOKEN,
      refresh_token: process.env.CISCOSPARK_REFRESH_TOKEN
    }
  },
  config: _config2.default
});

/**
 * While environment variables are handy for development, they don't really help
 * you write an app for lots of users. You can pass credentials to the spark
 * using init.
 * @param {Object} attrs
 * @param {Object} attrs.credentials
 * @memberof CiscoSpark
 * @returns {CiscoSpark}
 */
ciscospark.init = function init(attrs) {
  if ((0, _has2.default)(attrs, 'credentials.access_token') || (0, _has2.default)('credentials.refresh_token')) {
    attrs.credentials.authorization = {};
    ['access_token', 'token_type', 'expires', 'expires_in', 'refresh_token', 'refresh_token_expires', 'refresh_token_expires_in'].forEach(function (key) {
      (0, _set2.default)(attrs, 'credentials.authorization.' + key, (0, _get2.default)(attrs, 'credentials.' + key));
      (0, _deleteProperty2.default)(attrs.credentials, key);
    });
  }

  attrs.config = (0, _defaults2.default)(attrs.config, _config2.default);

  return new _sparkCore2.default(attrs);
};

exports.default = ciscospark;
