// @ciscospark/spark-core/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.grantErrors = exports.Credentials = exports.Authorization = exports.SparkTrackingIdInterceptor = exports.RequestTimingInterceptor = exports.RequestLoggerInterceptor = exports.ResponseLoggerInterceptor = exports.RedirectInterceptor = exports.registerPlugin = exports.NetworkTimingInterceptor = exports.AuthInterceptor = exports.SparkPlugin = exports.SparkHttpError = exports.default = exports.children = undefined;

var _sparkCore = require('./spark-core');

Object.defineProperty(exports, 'children', {
  enumerable: true,
  get: function get() {
    return _sparkCore.children;
  }
});
Object.defineProperty(exports, 'default', {
  enumerable: true,
  get: function get() {
    return _interopRequireDefault(_sparkCore).default;
  }
});
Object.defineProperty(exports, 'SparkHttpError', {
  enumerable: true,
  get: function get() {
    return _sparkCore.SparkHttpError;
  }
});
Object.defineProperty(exports, 'SparkPlugin', {
  enumerable: true,
  get: function get() {
    return _sparkCore.SparkPlugin;
  }
});
Object.defineProperty(exports, 'AuthInterceptor', {
  enumerable: true,
  get: function get() {
    return _sparkCore.AuthInterceptor;
  }
});
Object.defineProperty(exports, 'NetworkTimingInterceptor', {
  enumerable: true,
  get: function get() {
    return _sparkCore.NetworkTimingInterceptor;
  }
});
Object.defineProperty(exports, 'registerPlugin', {
  enumerable: true,
  get: function get() {
    return _sparkCore.registerPlugin;
  }
});
Object.defineProperty(exports, 'RedirectInterceptor', {
  enumerable: true,
  get: function get() {
    return _sparkCore.RedirectInterceptor;
  }
});
Object.defineProperty(exports, 'ResponseLoggerInterceptor', {
  enumerable: true,
  get: function get() {
    return _sparkCore.ResponseLoggerInterceptor;
  }
});
Object.defineProperty(exports, 'RequestLoggerInterceptor', {
  enumerable: true,
  get: function get() {
    return _sparkCore.RequestLoggerInterceptor;
  }
});
Object.defineProperty(exports, 'RequestTimingInterceptor', {
  enumerable: true,
  get: function get() {
    return _sparkCore.RequestTimingInterceptor;
  }
});
Object.defineProperty(exports, 'SparkTrackingIdInterceptor', {
  enumerable: true,
  get: function get() {
    return _sparkCore.SparkTrackingIdInterceptor;
  }
});

var _credentials = require('./plugins/credentials');

Object.defineProperty(exports, 'Authorization', {
  enumerable: true,
  get: function get() {
    return _credentials.Authorization;
  }
});
Object.defineProperty(exports, 'Credentials', {
  enumerable: true,
  get: function get() {
    return _credentials.Credentials;
  }
});
Object.defineProperty(exports, 'grantErrors', {
  enumerable: true,
  get: function get() {
    return _credentials.grantErrors;
  }
});

require('./plugins/logger');

require('./plugins/credentials');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
