// @ciscospark/spark-core/dist/interceptors/spark-tracking-id.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _defineProperties = require('babel-runtime/core-js/object/define-properties');

var _defineProperties2 = _interopRequireDefault(_defineProperties);

var _getPrototypeOf = require('babel-runtime/core-js/object/get-prototype-of');

var _getPrototypeOf2 = _interopRequireDefault(_getPrototypeOf);

var _classCallCheck2 = require('babel-runtime/helpers/classCallCheck');

var _classCallCheck3 = _interopRequireDefault(_classCallCheck2);

var _createClass2 = require('babel-runtime/helpers/createClass');

var _createClass3 = _interopRequireDefault(_createClass2);

var _possibleConstructorReturn2 = require('babel-runtime/helpers/possibleConstructorReturn');

var _possibleConstructorReturn3 = _interopRequireDefault(_possibleConstructorReturn2);

var _inherits2 = require('babel-runtime/helpers/inherits');

var _inherits3 = _interopRequireDefault(_inherits2);

var _httpCore = require('@ciscospark/http-core');

var _uuid = require('uuid');

var _uuid2 = _interopRequireDefault(_uuid);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @class
 */
/**!
 *
 * Copyright (c) 2015 Cisco Systems, Inc. See LICENSE file.
 * @private
 */

var SparkTrackingIdInterceptor = function (_Interceptor) {
  (0, _inherits3.default)(SparkTrackingIdInterceptor, _Interceptor);

  /**
   * @constructor
   * @param {ProxySpark} spark
   * @param {Object} options
   * @returns {SparkTrackingIdInterceptor}
   */

  function SparkTrackingIdInterceptor(spark, options) {
    (0, _classCallCheck3.default)(this, SparkTrackingIdInterceptor);

    var _this = (0, _possibleConstructorReturn3.default)(this, (0, _getPrototypeOf2.default)(SparkTrackingIdInterceptor).call(this, { spark: spark }));

    if (!options || !options.prefix) {
      throw new Error('`options.prefix` is required. Did you forget to specify it at config.trackingIdPrefix?');
    }

    var sequence = 0;
    var base = options.base || _uuid2.default.v4();
    var prefix = options.prefix || 'spark-js-sdk';

    (0, _defineProperties2.default)(_this, {
      base: {
        writable: false,
        value: base
      },
      prefix: {
        writable: false,
        value: prefix
      },
      sequence: {
        get: function get() {
          sequence += 1;
          return sequence;
        }
      }
    });
    return _this;
  }

  /**
   * @param {Object} options
   * @returns {SparkTrackingIdInterceptor}
   */


  (0, _createClass3.default)(SparkTrackingIdInterceptor, [{
    key: 'onRequest',


    /**
     * @see Interceptor#onRequest
     * @param {Object} options
     * @returns {Object}
     */
    value: function onRequest(options) {
      if (this.requiresTrackingId(options)) {
        options.headers.trackingid = this._generateTrackingId(options);
      }
      return options;
    }

    /**
     * Determines whether or not include a tracking id
     * @param {Object} options
     * @returns {boolean}
     */

  }, {
    key: 'requiresTrackingId',
    value: function requiresTrackingId(options) {
      return !options.headers.trackingid;
    }

    /**
     * Generates a tracking id
     * @private
     * @returns {string}
     */

  }, {
    key: '_generateTrackingId',
    value: function _generateTrackingId() {
      return this.prefix + '_' + this.base + '_' + this.sequence;
    }
  }], [{
    key: 'create',
    value: function create(options) {
      return new SparkTrackingIdInterceptor(this, options);
    }
  }]);
  return SparkTrackingIdInterceptor;
}(_httpCore.Interceptor);

exports.default = SparkTrackingIdInterceptor;
